src/pcre2test.c

```diff
@@ -8478,11 +8481,15 @@ the unused start of the buffer unaddressable. If we are using the POSIX
 interface, or testing zero-termination, we must include the terminating zero in
 the usable data. */
 
-c = code_unit_size * (((pat_patctl.control & CTL_POSIX) +
-                       (dat_datctl.control & CTL_ZERO_TERMINATE) != 0)? 1:0);
-pp = memmove(dbuffer + dbuffer_size - len - c, dbuffer, len + c);
-#ifdef SUPPORT_VALGRIND
-  VALGRIND_MAKE_MEM_NOACCESS(dbuffer, dbuffer_size - (len + c));
+c = code_unit_size * ((((pat_patctl.control & CTL_POSIX) != 0) +
+                       ((dat_datctl.control & CTL_ZERO_TERMINATE) != 0))? 1 : 0);
+
+#ifndef SUPPORT_VALGRIND
+pp = memmove(dbuffer + dbuffer_size - (len + c), dbuffer, len + c);
+#else
+pp = memmove(dbuffer + dbuffer_size - (len + c + x), dbuffer, len + c);
+VALGRIND_MAKE_MEM_NOACCESS(dbuffer, dbuffer_size - (len + c + x));
+VALGRIND_MAKE_MEM_NOACCESS(dbuffer + dbuffer_size - x, code_unit_size);
```

Comment:
You're adding an extra unit at the end of the buffer. But is this actually necessary? Can you give a test case where it makes a difference?

Valgrind will already warn (loudly) if you read past the end of the buffer, so extending the buffer and then telling it not to read the final bytes seems like extra complexity.

And - it makes it potentially _less_ likely for bugs to be caught. Before, a standard build without `SUPPORT_VALGRIND` would catch read-past-the-end errors (if run under valgrind), whereas now you actually have to specially build the library to make the VALGRIND_MAKE_MEM_NOACCESS calls in order for this to work.

Reply:
The extra unit was only added when building with `SUPPORT_VALGRIND` so the normal usage (including running with valgrind a build that didn't have that) was not affected, but agree that the extra complexity was unwarranted